### DISCUSSION TITLE
Some ANTLR  runtime 4.x bundles can expose newer ANTLR runtime 4.x content if installed.

### DISCUSSION THREAD

**Discussion by ylussaud** [eclipse-orbit/orbit-simrel#44] [General]
For instance if both ANTLR runtime 4.7.2 and ANTLR runtime 4.10.1 are installed in the same environment:
https://www.eclipse.org/downloads/download.php?file=/tools/orbit/downloads/drops/R20230302014618/repository/plugins/org.antlr.runtime.source_4.7.2.v20221112-0806.jar
https://www.eclipse.org/downloads/download.php?file=/tools/orbit/downloads/drops/R20230302014618/repository/plugins/org.antlr.runtime.source_4.10.1.v20230210-1433.jar

ANTLR runtime 4.7.2 exposes classes from ANTLR runtime 4.10.1. This leads to parsers generated with version 4.7.2 to raise an error when loaded even if the containing bundle has a dependency range that exclude other versions of ANTLR runtime.
The plugin dependency view shows the issue:

![image](https://github.com/user-attachments/assets/3194705e-68ce-4bb3-adb9-72bd049c9c38)

To fixe this one should remove the `Import-Package` from the manifest:
```
Import-Package: org.antlr.v4.gui;resolution:=optional,org.antlr.v4.runti
 me;resolution:=optional;version="[4.7,5)",org.antlr.v4.runtime.atn;reso
 lution:=optional;version="[4.7,5)",org.antlr.v4.runtime.dfa;resolution:
 =optional;version="[4.7,5)",org.antlr.v4.runtime.misc;resolution:=optio
 nal;version="[4.7,5)",org.antlr.v4.runtime.tree;resolution:=optional;ve
 rsion="[4.7,5)",org.antlr.v4.runtime.tree.pattern;resolution:=optional;
 version="[4.7,5)",org.antlr.v4.runtime.tree.xpath;resolution:=optional;
 version="[4.7,5)"
```

Other versions might also be affected... Would it be possible to contribute patched versions to Orbit ?

**Comment by merks** [comment]
Orbit currently provides these versions:

https://download.eclipse.org/tools/orbit/simrel/orbit-aggregation/table.html

![image](https://github.com/user-attachments/assets/2120953f-23d0-494e-925b-9ab1f227a46c)

The 4.x version is consumed directly from Maven central where it is available as an OSGi build:

https://repo1.maven.org/maven2/org/antlr/antlr4-runtime/4.13.2/

Projects should be using these versions only.   It appears that Acceleo is ignoring that.  That's not a problem for Orbit to solve.  

I assume in the past that this was used and that Orbit rewrapped it:

https://repo1.maven.org/maven2/org/antlr/antlr4-runtime/4.7.2/

But the original artifact did not have that problem:
```
Manifest-Version: 1.0
Export-Package: org.antlr.v4.runtime;uses:="org.antlr.v4.runtime.atn,o
 rg.antlr.v4.runtime.dfa,org.antlr.v4.runtime.misc,org.antlr.v4.runtim
 e.tree,org.antlr.v4.runtime.tree.pattern";version="4.7.2",org.antlr.v
 4.runtime.misc;uses:="org.antlr.v4.runtime";version="4.7.2",org.antlr
 .v4.runtime.tree;uses:="org.antlr.v4.runtime,org.antlr.v4.runtime.mis
 c";version="4.7.2",org.antlr.v4.runtime.tree.xpath;uses:="org.antlr.v
 4.runtime,org.antlr.v4.runtime.atn,org.antlr.v4.runtime.tree";version
 ="4.7.2",org.antlr.v4.runtime.tree.pattern;uses:="org.antlr.v4.runtim
 e,org.antlr.v4.runtime.misc,org.antlr.v4.runtime.tree";version="4.7.2
 ",org.antlr.v4.runtime.dfa;uses:="org.antlr.v4.runtime,org.antlr.v4.r
 untime.atn";version="4.7.2",org.antlr.v4.runtime.atn;uses:="org.antlr
 .v4.runtime,org.antlr.v4.runtime.dfa,org.antlr.v4.runtime.misc";versi
 on="4.7.2"
Implementation-Title: ANTLR 4 Runtime
Implementation-Version: 4.7.2
Built-By: parrt
Tool: Bnd-2.4.1.201501161923
Bundle-Name: ANTLR 4 Runtime
Created-By: Apache Maven Bundle Plugin
Implementation-URL: http://www.antlr.org/runtime/antlr4-runtime
Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.7))"
Bundle-Vendor: ANTLR
Implementation-Vendor: ANTLR
Automatic-Module-Name: org.antlr.antlr4.runtime
Implementation-Vendor-Id: org.antlr
Bundle-Version: 4.7.2
Build-Jdk: 1.7.0_65
Bnd-LastModified: 1545158213996
Bundle-ManifestVersion: 2
Bundle-License: http://www.antlr.org/license.html
Bundle-Description: The ANTLR 4 Runtime
Bundle-SymbolicName: org.antlr.antlr4-runtime
Bundle-DocURL: http://www.antlr.org
Archiver-Version: Plexus Archiver
```

The Acceleo project (Obeo) should use what Orbit provides.  It's no longer possible in the latest Orbit design to re-wrap bundles so this problem cannot be fixed by Obit.  Sorry.

(I appear to me that those folks simply need to use the newer bundle symbolic name. )

**Comment by ylussaud** [comment]
Thank you for your answer, I'll check if we can update our ANTLR dependencies.
